server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1311,37 +1312,47 @@ void run(ActionListener<Void> listener) {
 
             private void determineShardCount(ActionListener<Void> listener) {
                 try (var listeners = new RefCountingListener(listener)) {
-                    for (final var indexMetaGeneration : snapshotIds.stream()
+                    Map<String, String> blobIdToIndexUuidMap = originalRepositoryData.indexMetaDataGenerations().getBlobIdToIndexUuidMap();
```

Comment:
This gets called once per index, which could still be (hundreds-of-)thousands of times, but it returns the same value each time. It really needs to be done once for the entire snapshot deletion process.